Diff/Diff.sublime-syntax

```diff
@@ -1,75 +1,402 @@
 %YAML 1.2
 ---
-# http://www.sublimetext.com/docs/3/syntax.html
+# https://www.sublimetext.com/docs/syntax.html
+# https://www.gnu.org/software/diffutils/manual/diffutils.html
 name: Diff
+scope: source.diff
+version: 2
+
 file_extensions:
   - diff
   - patch
+
 first_line_match: |-
-  (?x)^
-      (===\ modified\ file
-      |==== \s* // .+ \s - \s .+ \s+ ====
-      |Index:[ ]
-      |---\ [^%]
-      |\*\*\*.*\d{4}\s*$
-      |\d+(,\d+)* (a|d|c) \d+(,\d+)* $
-      |diff\ --git[ ]
-      )
+  ^(?x:
+    ===\ modified\ file
+  | ==== \s* // .+ \s - \s .+ \s+ ====
+  | Index:[ ]
+  | Prereq:[ ]
+  | ---\ [^%]
+  | \*\*\*.*\d{4}\s*$
+  | \d+(,\d+)* [adc] \d+(,\d+)* $
+  )
 
-scope: source.diff
 contexts:
   main:
-    - match: '^((\*{15})|(={67})|(-{3}))$\n?'
+    - match: ^(?={{git_first_line}})
+      push: Packages/Git Formats/Git Diff.sublime-syntax#email-first-line
+    - include: leading-text
+    - include: line-ranges
+    - include: headers
+    - include: deltas
+    - include: diff3-normal
```

Comment:
As _Git Diff_ extends from this syntax all these contexts are compiled into _Git Diff_ as well even though they seem unused. It shouldn't cause any performance issues as they are unreferenced in _Git Diff_ but they cause larger syntax cache file.

Hence I wonder how big overlap between _Diff_ and _Git Diff_ really is. I just found a couple of small contexts on a first glance.

Depending on the answer I'd suggest to ...

1. either decouble syntaxes completely, by duplicating those patterns - if there are only few
2. or create a common base syntax, which only contains shared contexts.


So with the comment above, _Diff_ and _Git Diff_ are basically compiled to the same binaries more or less, sharing all their contexts.

That's somehow odd, to be honest.

Reply:
Thanks for pointing this out. There are a couple ways to handle it that I've thought about (and perhaps more that I hadn't considered). Please feel free to comment.

### Merge Git Diff back into Diff

There's not really a big drawback to this. I originally had them both in the same file. I split them because the Git part started feeling unwieldy compared to the nice, tight Diff part.

### Decouple completely
*Feels* unsatisfying, but there aren't really any practical drawbacks. (Possible exception: Does the Diff "match-Git-first-line / push Git Diff" end up duplicating all those contexts back anyway?)

### Base syntax (simple)
Just centralize the unified diff stuff into a base syntax.

### Base syntax (split)
 Create a set of very small Diff syntaxes each handling their specific case, and then mash them all together for a generic Diff syntax.

- Diff (common)
    - line-marked diffs
        - Diff Normal
        - Diff Context
        - Unified Diff
            - Git Diff
    - section-marked diffs
        - Diff3 Normal
        - Diff3 Edit
        - Conflict (?)
    - ??

With a main Diff syntax that extends everyone else's for convenience. 

#### Pros
- You have the opportunity to rigidly highlight as a particular format if you choose to.
#### Cons
- It's a lot of tiny files that *probably* no one will care to use.
- We may get to find out what happens when one syntax prepends to a context, another one replaces it completely, and a third one inherits from both.

Reply:
Last option indeed feels a bit like overkill with regards to amount of files and scattering contexts. If all those patterns don't conflict it is ok to have them in one file.

How much would a common _Diff (Basic)_ syntax contain? If it is enough, it could be a base for various dialects like _SQL (Basic)_ or _RegExp (Basic)_. Maybe most cleanest solution.

If it however contains only 2 - 6 tiny contexts, it might probably also ok to just drop normal Diff contexts in _Git Diff_ by setting them empty via e.g. `diff3-normal: []`. This drops unused contexts but looks a bit messy if there are too many of them.

Reply:
What do you think about https://github.com/michaelblyons/SublimeSyntax-Defaults/pull/7 ? I tried making a "Diff (Basic)," but there were some dependency chains that made it kind of ugly. This way is much cleaner.

Reply:
I wouldn't primarily distinguish between unified as base and the rest, but just put common stuff shared accross GNU Diff and Git Diff into a base syntax.

That said, I take on your approach by removing dedundants and renaming contexts, preparing a suggestion for an (hopefully) optimized extending architecture. I have a feeling of it possibly resolving some of current weirdnesses.

Reply:
Sorry I didn't notice your comment before I did some force-pushing. Hopefully not too disruptive. :sweat_smile:

Reply:
Thanks for your help! The suggested changes look great.